### PR TITLE
HW_Parallel_programming

### PR DESCRIPTION
**Pull request by UsernameAlexei**
HW_Parallel_programming

### CODE REVIEW

File: download.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    parser.add_argument('-t', '--threads', default=4, nargs='?', type=int)
+    parser.add_argument('-s', '--size', default='128x128', nargs='?')
+
+    return parser
+
+
+# Загрузка файлов
+def download_file(url, name):
+    """Переменные для статистики"""
+    global DOWNLOAD_COMPLETE, DOWNLOAD_ERROR
```
Comment (agorlan) [inline download.py]: I suppose we need locks in order to use these global variables from different threads

File: download.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+    """Скачиваем файл"""
+    fname = name
+    try:
+        handle = urllib.request.urlopen(url)
+        filepath = os.path.join(dirname, fname)
+
+        with open(filepath, "wb") as f:
+            print(f'download file {fname} complete')
+            DOWNLOAD_COMPLETE += 1
```
Comment (agorlan) [inline download.py]: Let's increase after all has been written

File: download.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+
+parser = createParser()
+namespace = parser.parse_args(sys.argv[1:])
+
+q = queue.Queue()
+
+# Проверяем/создаем директорию
+try:
+    file = open(namespace.urls_file)
```
Comment (agorlan) [inline download.py]: Use `with` in order to close file after content is read

File: download.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+start_time = datetime.now()
+
+
+def createParser():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('urls_file', default='urllist.txt', nargs='?')
+    parser.add_argument('-d', '--dir', default='thumbnails/', nargs='?', type=str)
+    parser.add_argument('-t', '--threads', default=4, nargs='?', type=int)
+    parser.add_argument('-s', '--size', default='128x128', nargs='?')
```
Comment (agorlan) [inline download.py]: Consider using custom type here with validation code

File: download.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+# Редактируем файлы
+with ThreadPoolExecutor(max_workers=namespace.threads) as executor:
+    x = executor.submit(resize)
+
+# Расчет общего количества байт
+root_directory = Path('thumbnails')
+sum_byte = sum(f.stat().st_size for f in root_directory.glob('**/*') if f.is_file())
+
+print(f'Number of downloaded files: {DOWNLOAD_COMPLETE + DOWNLOAD_ERROR}\n'
+      f'Number of downloaded bytes: {sum_byte}\n'
```
Comment (agorlan) [inline download.py]: For some reason I get 
"Number of downloaded bytes: 0"
for any run (ubuntu)

File: download.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+DOWNLOAD_ERROR = 0
+with ThreadPoolExecutor(max_workers=namespace.threads) as executor:
+    f = executor.map(download_file, url, name)
+
+path = namespace.dir
+dirs = os.listdir(path)
+
+# Редактируем файлы
+with ThreadPoolExecutor(max_workers=namespace.threads) as executor:
+    x = executor.submit(resize)
```
Comment (agorlan) [inline download.py]: Let's make parallel resize. For now one executor works with all files. We can refactor resize to handle one file and call submit for all files

File: download.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+# Проверяем/создаем директорию
+try:
+    file = open(namespace.urls_file)
+    url = [x.strip() for x in file]
+
+except FileNotFoundError:
+    print(f'File is missing')
+
+name = [f'{x}.jpg' for x in range(len(url))]
```
Comment (agorlan) [inline download.py]: file_names

File: download.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+dir_name = namespace.dir
+
+for i in url:
+    q.put(i)
```
Comment (agorlan) [inline download.py]: If we do not get anything from queue then putting it there makes no sense

File: download.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+path = namespace.dir
+dirs = os.listdir(path)
+
+# Редактируем файлы
+with ThreadPoolExecutor(max_workers=namespace.threads) as executor:
+    x = executor.submit(resize)
+
+# Расчет общего количества байт
+root_directory = Path('thumbnails')
```
Comment (agorlan) [inline download.py]: Hardcoded folder name. Works only when custom output folder is not provided

File: download.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    global DOWNLOAD_COMPLETE, DOWNLOAD_ERROR
+    """Создаем папку"""
+    dirname = namespace.dir
+    if not os.path.isdir(namespace.dir):
+        os.mkdir(namespace.dir)
+
+    """Скачиваем файл"""
+    fname = name
+    try:
+        handle = urllib.request.urlopen(url)
```
Comment (agorlan) [inline download.py:72]: Please, use handle.length in order to count bytes downloaded

File: download.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+path = namespace.dir
+dirs = os.listdir(path)
+
+# Редактируем файлы
+with ThreadPoolExecutor(max_workers=namespace.threads) as executor:
+    x = executor.submit(resize)
+
+# Расчет общего количества байт
+root_directory = Path('thumbnails')
+sum_byte = sum(f.stat().st_size for f in root_directory.glob('**/*') if f.is_file())
```
Comment (agorlan) [inline download.py]: Dangerous. If folder was not empty, we count files which was laying there before we downloaded anything. Use length property of object returned by urlopen()

File: download.py (hunk 1 above)
Comment (UsernameAlexei) [reply download.py]: Yes, added

File: download.py (hunk 2 above)
Comment (UsernameAlexei) [reply download.py]: Updated

File: download.py (hunk 3 above)
Comment (UsernameAlexei) [reply download.py]: Updated

File: download.py (hunk 4 above)
Comment (UsernameAlexei) [reply download.py]: Added custom type

File: download.py (hunk 6 above)
Comment (UsernameAlexei) [reply download.py]: Updated

File: download.py (hunk 7 above)
Comment (UsernameAlexei) [reply download.py]: Updated

File: download.py (hunk 9 above)
Comment (UsernameAlexei) [reply download.py]: Update. Forgot to fix this part right away

File: download.py (hunk 10 above)
Comment (UsernameAlexei) [reply download.py:72]: Updated
Maybe so?
```python
    try:
        handle = urllib.request.urlopen(url)
        filepath = os.path.join(dirname, fname)

        length_in_bytes = os.stat(filepath).st_size
        SUM_BYTE += length_in_bytes
```

File: download.py (hunk 11 above)
Comment (UsernameAlexei) [reply download.py]: Yes indeed. Updated

File: download.py (hunk 5 above)
Comment (UsernameAlexei) [reply download.py]: Update

**Comment by UsernameAlexei** [comment]
Andrey, thanks for the comments. Sorry for the delay with corrections. 
Made changes. The code looks better.
The module is interesting.

**Pull request merged**